### PR TITLE
improve index on appconfig

### LINKED ISSUE #43103: [Bug]: Redundant index appconfig.ac_app_lazy_i
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

```sql
create index ac_app_lazy_i
    on nextcloud.oc_appconfig (appid, lazy);

create index ac_app_lazy_key_i
    on nextcloud.oc_appconfig (appid, lazy, configkey);

```

the concatenated index ac_app_lazy_key_i fully covers the two first columns of ac_app_lazy_i. The database can also use ac_app_lazy_key_i when only the two first columns are accessed. ac_app_lazy_i is redundant. And hurts performance.

### Steps to reproduce

1. Install Nextcloud

### Expected behavior

Efficient use of database indices.

### Installation method

None

### Nextcloud Server version

master

### Operating system

None

### PHP engine version

None

### Web server

None

### Database engine version

None

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

_No response_

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

Introduced with https://github.com/nextcloud/server/pull/41755.

:eyes: @ArtificialOwl and @artonge @come-nc @nickvergessen as reviewers

### PR DESCRIPTION
**Pull request by ArtificialOwl**
follow-up of #42820

- [x] only create new fields if not existing yet
- [x] removing `appconfig_configkey_key` index on single field configkey as there is no search on this field
- [x] redefine `lazy` as `Types::INTEGER` (smallint(6)) instead of `Types::BOOLEAN` to be able to set it a `NotNull` (set in 28-rc2)
- [x] index on `lazy`
- [x] removing the use of `$migrationCompleted` that was needed during the migration from 28 to 29 but not anymore since we generate the (previously) missing fields `lazy` and `type` in 28
- [x] do not select `lazy` during loadConfig as we already know its value

_Note: After some discussion, it seems better to switch `lazy` to INTEGER so the field can be set as NotNull (because Oracle)_

Fixes https://github.com/nextcloud/server/issues/43103

### CODE REVIEW

**Review by artonge — APPROVED** [review]

File: core/Migrations/Version29000Date20231126110901.php [hunk 1]
```diff
@@ -43,23 +43,29 @@ public function changeSchema(IOutput $output, Closure $schemaClosure, array $opt
 			return null;
 		}
 
+		$changed = false;
```
Comment (nickvergessen) [inline core/Migrations/Version29000Date20231126110901.php:46]: this migration executed already. You can not queue another one before it neither should you modify it.
Please generate 2 new migrations. If the changed from this here should be skipped, you can comment out the code, but you need another migration making sure the stuff is removed and then another migration to add the things again with the correct type.

Comment (ArtificialOwl) [reply core/Migrations/Version29000Date20231126110901.php:46]: I have created a [core/Migrations/Version29000Date20231126110900.php](https://github.com/nextcloud/server/pull/42903/files#diff-679f3baf9ad965a15ccb3ed83f05b6230389dd9f70b0fa7c4b6e650cb5957eee) that drops erroneous field `lazy` so we can switch to integer

File: core/Migrations/Version28000Date20231126110901.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+use Closure;
+use OCP\DB\ISchemaWrapper;
+use OCP\DB\Types;
+use OCP\Migration\IOutput;
+use OCP\Migration\SimpleMigrationStep;
+
+/**
+ * Create new fields for type and lazy loading in appconfig for the new IAppConfig API.
+ */
+class Version28000Date20231126110901 extends SimpleMigrationStep {
```
Comment (nickvergessen) [inline core/Migrations/Version28000Date20231126110901.php:37]:
```suggestion
class Version29000Date20240123... extends SimpleMigrationStep {
```
Please generate a proper new migration that executes **after** the already existing one, so there is a proper stream line of migrations and an easy traceable DB state of instanced running dailies or git.

File: core/Migrations/Version28000Date20231126110901.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	public function changeSchema(IOutput $output, Closure $schemaClosure, array $options): ?ISchemaWrapper {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+
+		if (!$schema->hasTable('appconfig')) {
+			return null;
+		}
```
Comment (nickvergessen) [inline core/Migrations/Version28000Date20231126110901.php:44]:
```suggestion
```
Don't check against pre requirements.

File: core/Migrations/Version29000Date20231126110900.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+use Closure;
+use OCP\DB\ISchemaWrapper;
+use OCP\Migration\IOutput;
+use OCP\Migration\SimpleMigrationStep;
+
+/**
+ * Create new fields for type and lazy loading in appconfig for the new IAppConfig API.
+ */
+class Version29000Date20231126110900 extends SimpleMigrationStep {
```
Comment (nickvergessen) [inline core/Migrations/Version29000Date20231126110900.php:36]:
```suggestion
class Version29000Date20240123... extends SimpleMigrationStep {
```
As above, this should have a later version and timestamp to have a linear DB state

File: core/Migrations/Version29000Date20231126110901.php (hunk 1 above)
Comment (nickvergessen) [reply core/Migrations/Version29000Date20231126110901.php:46]: The changes in this file should still be reverted. You can comment out the lazy field and wrong indices, but only comment out, not do any code level changes.

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: core/Migrations/Version28000Date20231126110901.php (hunk 2 above)
Comment (nickvergessen) [reply core/Migrations/Version28000Date20231126110901.php:37]: Okay I see only now that you changed the version number in the backport. That should not have happened.

Anyway, we should still have the new and fixing migration with a 2401... Timestamp at the end of the chain and the 28...2311 migrations both unchanged from previous PRs.

**Comment by ArtificialOwl** [comment]
I have estimated that a full squash+rebase was a lot better as I reverted a lot of code:

- upstream of #42820 to master
- commenting most of the code from `Version28000Date20231126110901` upstream of #42820 to master (+ explication)
- commenting most of the code of previous migration step `Version28000Date20231126110901 ` (+ explication)
- new migration step `Version29000Date20240124132201` to remove previous boolean `lazy` field and related indexes, fix `type` to **unsigned** and remove useless index on `configkey`.
- new migration step `Version29000Date20240124132202` to create a **smallint** `lazy` field.
- keep migration fall back on reading app config values to keep upgrade compatibility with versions prior to 28.0.2
- switching write/read request to `lazy` field from managing a **boolean** to managing a **smallint**
- remove `lazy` from select list if known when calling `loadConfig()`

File: core/Migrations/Version29000Date20240124132201.php [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+namespace OC\Core\Migrations;
+
+use Closure;
+use OCP\DB\ISchemaWrapper;
+use OCP\DB\Types;
+use OCP\Migration\IOutput;
+use OCP\Migration\SimpleMigrationStep;
+
+/**
+ * Create new fields for type and lazy loading in appconfig for the new IAppConfig API.
```
Comment (nickvergessen) [inline core/Migrations/Version29000Date20240124132201.php:35]:
```suggestion
 * Create new column for type and remove previous lazy column in appconfig (will be recreated by Version29000Date20240124132202) for the new IAppConfig API.
```

File: core/Migrations/Version29000Date20240124132202.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+namespace OC\Core\Migrations;
+
+use Closure;
+use OCP\DB\ISchemaWrapper;
+use OCP\DB\Types;
+use OCP\Migration\IOutput;
+use OCP\Migration\SimpleMigrationStep;
+
+/**
+ * Create new fields for type and lazy loading in appconfig for the new IAppConfig API.
```
Comment (nickvergessen) [inline core/Migrations/Version29000Date20240124132202.php:35]:
```suggestion
 * Create new column and index for lazy loading in appconfig for the new IAppConfig API.
```

**Review by nickvergessen — APPROVED** [review]

**Pull request merged**